Arm64 vector ABI

CarolEidt opened the pull request:
Extend HFA support to support vectors as well as floating point types.
This requires that the JIT recognize vector types even during crossgen,
so that the ABI is supported consistently.

Fix #16022

CarolEidt commented:
@dotnet/jit-contrib @tannergooding @sdmaclea @tamarchristinaarm @jkotas @jkoritzinsky 
This is not yet ready to merge (in fact, there's a conflict with @jkoritzinsky 's recent commit), but I would like to get feedback at this point.

mikedn commented on src/jit/compiler.h:
```diff
@@ -682,54 +680,32 @@ class LclVarDsc
+        assert(slots <= 8);
+#elif defined(_TARGET_ARM64_)
+        return lvExactSize >> (2 + _lvHfaLogSize);
+        assert(slots <= 4);
```
This is odd, why is there an assert after a return? And `slots` is 0 anyway. I'd guess that the return was actually supposed to be a `slots` assignment.

CarolEidt replied:
Yes, indeed. Thanks.

mikedn commented on src/jit/compiler.h:
```diff
@@ -7612,6 +7612,18 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
+    // Should we recognize SIMD types?
+    // We always do this on ARM64 to support HVA types.
+    __declspec(property(get = GetSupportSIMDTypes)) bool supportSIMDTypes;
```
I don't think new `__declspec(property)` should be added to the JIT...

CarolEidt replied:
Yes, I suppose not. Believe it or not, I've been working on this for so long that I had added this before the (relatively) recent discussion of removing `__declspec`s.

mikedn replied:
Ha ha, I believe you. Sadly I don't think it's currently possible to do changes in the ABI area of the JIT without a significant amount of time and care.

CarolEidt replied:
> Sadly I don't think it's currently possible to do changes in the ABI area of the JIT without a significant amount of time and care

Right, plus, for this change I had to delve into the VM side, which I'm much less familiar with.

mikedn commented on src/jit/vartype.h:
```diff
@@ -276,6 +276,128 @@ inline bool varTypeIsEnregisterableStruct(T vt)
+inline unsigned char logSizeForHfaType(T vt)
+{
+    unsigned char logSize = 0;
+    switch (vt)
```
Here you probably want `switch (TypeGet(vt))` if you want the template to work as intended. Though I question the need for such templates...

CarolEidt replied:
Yes, thanks! I don't have a strong opinion either way on the use of these templates, but since this seemed to belong in vartype.h, I went with the prevailing style there.

mikedn commented on src/jit/vartype.h:
```diff
@@ -176,7 +176,7 @@ inline bool varTypeIsI(T vt)
 inline bool varTypeCanReg(T vt)
 {
-    return ((varTypeClassification[TypeGet(vt)] & (VTF_INT | VTF_I | VTF_FLT)) != 0);
+    return (TypeGet(vt) != TYP_STRUCT);
```
It looks like this function has became a duplicate of `varTypeIsEnregisterableStruct`.

CarolEidt replied:
Yes, I hadn't realized they were duplicates. I'll look into eliminating this one (I prefer the more verbose naming, though could be convinced otherwise).

mikedn replied:
Verbose or not it seems to me that both names have some issues. `varTypeCanReg` doesn't sound quite right to me from the English language point of view (varTypeCanRegister or what?). The other one has a "Struct" suffix that seems redundant/misleading, might be nice to remove that if it's not too much churn.

briansull replied:
maybe just drop the struct and call it:

`varTypeIsEnregisterable`

CarolEidt replied:
Yes, I've just done that. There are not so many uses that it would create a lot of churn.

mikedn commented on src/jit/compiler.h:
```diff
@@ -1487,8 +1471,8 @@ struct fgArgTabEntry
-    bool _isHfaArg : 1;    // True when the argument is an HFA type.
-    bool _isDoubleHfa : 1; // True when the argument is an HFA, with an element type of DOUBLE.
+    bool          _isHfaArg : 1;   // True when the argument is an HFA type.
+    unsigned char _hfaLogSize : 2; // The base 2 log of the HFA element size in 4-byte units (e.g. 0 for TYP_FLOAT)
```
Isn't there other terminology that could be used here? "size" most of the type refers to byte size. In call morphing code "size" also refers to the number of slots. And now here it refers to the number of 4-byte units.

CarolEidt replied:
How about `_hfaLogWidth`? Or "_hfaLogRelSize" for relative size?

mikedn replied:
Hmm, width is also used in some cases to denote byte size (some block ops code does that). But at least it's not used within `LclVarDsc` (which has size and exactSize) so it's a better alternative.

Though I'm not sure why this is needed to begin with. Is it just to "compress" the information down to 2 bits in the bitfield?

CarolEidt replied:
Yes, it's just to keep the information small as opposed to making it a `var_types` field, or creating a new enum.

mikedn replied:
Hmm, then leave it as it is. I have a feeling that this HFA info will end up in the `ClassLayout` thing I have in #21705 sooner or later.

CarolEidt replied:
In looking at my response, I'm thinking a new enum would be better - it could be called something like `HfaElemKind` and have values `HFA_ELEM_FLOAT`, `HFA_ELEM_DOUBLE` and `HFA_ELEM_VECTOR128`. And then we could rely on the C++ compiler to generate good code rather than trying to pre-optimize by using shifts. (I try to make it a principle to rely on the code generator and not pre-optimize, but I don't always manage to do so!)

mikedn replied:
Right, an enum should be fine as long as it has underlying type `unsigned char` to match the other bitfield members. Otherwise AFAIR VC++ does not pack data member of different types into the same byte of the bitfield.

Mistuke commented:
Ah great, I'm out of office until Monday but will take a look first thing Monday morning!

TamarChristinaArm commented:
Ah great, I'm out of office until Monday but will take a look first thing Monday morning!

AndyAyersMS commented on src/jit/codegencommon.cpp:
```diff
@@ -10160,7 +10160,11 @@ bool Compiler::IsMultiRegReturnedType(CORINFO_CLASS_HANDLE hClass)
 
+#ifdef _TARGET_ARM64_
+    return (varTypeIsStruct(returnType) && (howToReturnStruct != SPK_PrimitiveType));
+#else
```
Probably also need to look for `SPK_EnclosingType`?

CarolEidt replied:
I don't think so, at least not for now. For the `SPK_EnclosingType` case the `returnType` will be `TYP_INT` or `TYP_LONG`, so we won't pass the first test.

CarolEidt commented on src/jit/codegencommon.cpp:
```diff
@@ -10206,7 +10206,20 @@ var_types Compiler::GetHfaType(CORINFO_CLASS_HANDLE hClass)
+            // HVAs are only supported on ARM64, and only for sizes of 8 or 16 bytes.
+            // For 8-byte vectors corType will be returned as CORINFO_TYPE_DOUBLE.
+            result = TYP_SIMD16;
+            JITDUMP("Found an HVA of SIMD16\n");
```
I'll remove it. When I first started working on this, we weren't dumping HFA info on locals or args. Now that I've added that info there, this isn't as useful.

CarolEidt commented:
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test

CarolEidt commented:
I believe I've incorporated all the feedback, and this is now green. I'm going to do some outerloop testing, but would appreciate any additional feedback on the changes.

CarolEidt commented:
/azp run coreclr-outerloop

sdmaclea commented:
Are there any tests which are testing this is working correctly?  Last I knew the tests were missing.
Are we supporting interop with external libraries using HVA? Especially this class of tests.

CarolEidt commented:
> Are there any tests which are testing this is working correctly? Last I knew the tests were missing.
Are we supporting interop with external libraries using HVA? Especially this class of tests.

We *should* be supporting interop with external libraries using HVA, though I don't know of a good way to test that now because we don't support interop with generic types (e.g. `Vector128<T>`). We have a test that tests managed-to-managed calls (JIT/Directed/VectorABI) but I haven't come up with a great way other than manual inspection to validate it.

sdmaclea commented:
@CarolEidt These tests were disable because the Vector ABI was broken

https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/HardwareIntrinsics/Arm64/Simd.cs#L1358-L1475
https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/HardwareIntrinsics/Arm64/Simd.cs#L1531-L1568

You may want to try reenabling them (possibly as part of this PR).

CarolEidt commented:
@sdmaclea - I gather that the ABI was broken for vectors independent of the change to pass them in registers. They indeed still fail, but the initial failure was due to an error in the (non)expansion of out-of-range immediates. Fixing that, it dies a horrible death presumably in the pinvoke stub. I'm tracking it down, but my limited knowledge of that area makes it slow going.
FYI @RussKeldorph - this is at even greater risk for 3.0 than I thought.

sdmaclea commented:
>  I gather that the ABI was broken for vectors independent of the change to pass them in registers.

Yes that is true. The SIMD function calls were not working either way.  My initial Vector ABI change was motivated to fix these tests. From my recollection, I believe it had succeeded.  But my fix was too large and complicated and too close to ZBB to merge.  It bit rotted too much for me to resurrect given I was moving off arm64 .NET.

> in the pinvoke stub

Hmmm? I thought it was supposed to jump to a switch table.  I guess the out of range case was supposed to throw? I guess that is the pinvoke stub.

sdmaclea commented:
@CarolEidt My quick review of the code.
```C++
//    The importer is responsible for inserting an upstream range check
//    (GT_HW_INTRINSIC_CHK) for swReg, so no range check is needed here.
//
template <typename HWIntrinsicSwitchCaseBody>
void CodeGen::genHWIntrinsicSwitchTable(regNumber                 swReg,
                                        regNumber                 tmpReg,
                                        int                       swMax,
                                        HWIntrinsicSwitchCaseBody emitSwCase)
```
GT_HW_INTRINSIC_CHK is calling `void CodeGen::genRangeCheck(GenTree* oper)` 

looks like `void CodeGen::genRangeCheck(GenTree* oper)` is not properly range checking the intrinsics.

CarolEidt commented:
@sdmaclea - actually I was mistaken. Things were looking a bit out of order, so it appeared that it was dying in a pinvoke stub, but actually it was dying in the JIT. I think I have a fix now.

CarolEidt commented:
> looks like void CodeGen::genRangeCheck(GenTree* oper) is not properly range checking the intrinsics

What's supposed to happen is that the range check is only inserted in the `mustExpand` case, so in the case where `mustExpand` is false, we must generate an actual call if we don't have an immediate or if it is out of range. Then the range check is done when compiling the recursive (`mustExpand`) case.

sdmaclea commented:
@CarolEidt Looks like your fix is reasonable.  Your description seems correct, and I remember verifying it worked correctly.  Either 
* I had a change to fix this and it was lost
* Code refactoring broke this because the relevant tests were disabled.

Thanks for fixing and reenabling the tests.

CarolEidt commented:
/azp run coreclr-outerloop

TamarChristinaArm commented:
I also can't spot anything that looks out of place ABI wise. From the couple of examples I have been running the new codegen looks correct and is a nice improvement over master! Thanks @CarolEidt

CarolEidt commented:
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test

CarolEidt commented:
/azp run coreclr-outerloop

sdmaclea commented:
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test

sdmaclea commented:
Apparently I can't trigger Windows arm builds...  It should be retriggered now that the fix has merged.

CarolEidt commented:
> Apparently I can't trigger Windows arm builds... It should be retriggered now that the fix has merged.

I don't think that the CI system will merge or rebuild at this point. But I'm going to re-launch (it got an I/O error last time it was re-launched) and if we get that same single failure that you fixed, I'll feel free to ignore it.

CarolEidt commented:
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test

BruceForstall commented:
> Apparently I can't trigger Windows arm builds

@sdmaclea That's because you removed yourself from the whitelist: https://github.com/dotnet/coreclr/pull/22924

sdmaclea commented:
@BruceForstall I am surprised the `arm64Users` whitelist affects `Windows arm` triggering. If the auto triggers have been removed, I can add myself back to the whitelist.  The side effect of having all my PRs red because arm64/arm was broken, motivated me to remove myself.

BruceForstall commented:
`arm64Users` controls access to the Windows arm64 machines, which also run Windows arm32 testing via WoW.

While PR testing obviously fluctuates in stability, and not just Windows arm32/arm64, hopefully failure is always a temporary state.

CarolEidt commented:
@dotnet/jit-contrib @tannergooding @sdmaclea @TamarChristinaArm @jkotas @jkoritzinsky this is ready to merge. The remaining failures are:
```
_JitBlue_GitHub_23791_GitHub_23791_GitHub_23791_sh
_Enum_hasflag_hasflag_sh
```
Which are new tests, for which the fix post-dates this PR.

CarolEidt commented:
@dotnet/jit-contrib ping

BruceForstall commented:
It seems that this change might warrant some addition to the CLR ABI document, describing our support for HVA. If it matches the ARM64 ABI definition precisely, then we would defer to that. But maybe it matters to define precisely what constitutes an HVA for .NET Core, how interop works, etc.?

https://github.com/dotnet/coreclr/blob/master/Documentation/botr/clr-abi.md

CarolEidt commented:
@BruceForstall - I agree that it would be good to update the CLR ABI document, though it is my belief that this matches the ARM64 ABI precisely - and that the types `Vector<T>`, `Vector64<T>` and `Vector128<T>` will be treated as vector types. Homogeneous aggregates of these will be passed as HVAs. The fixed vector types (`Vector2`, `Vector3`, and `Vector4`) are explicitly structs of floats, so they will continue to be passed as HFAs of float.
It would still probably merit an update of the document, to clarify that (for now) we don't support interop because generic types are not supported.
That said, I'd prefer to merge this first, as I think it may take some time to update the document, agree on wording, etc. and the likelihood of conflicts will increase over time.

tannergooding commented on src/jit/codegencommon.cpp:
```diff
@@ -10204,7 +10204,19 @@ var_types Compiler::GetHfaType(CORINFO_CLASS_HANDLE hClass)
+        if (corType == CORINFO_TYPE_VALUECLASS)
+        {
+            // This is a vector type.
+            // HVAs are only supported on ARM64, and only for sizes of 8 or 16 bytes.
```
Could you clarify on the "only for sizes of 8 or 16 bytes"?

The spec says:
> An Homogeneous Short-Vector Aggregate (HVA) is an Homogeneous Aggregate with a Fundamental Data Type that is a Short-Vector type and at most four uniquely addressable members.

A `short-vector` is then classified as a `64-bit` or `128-bit` vector.

CarolEidt replied:
I'm not sure what clarification you're looking for. It seems like the comment and the spec are saying the same thing. 
Oh - by "sizes" I meant "vector sizes". Is that what needs clarifying?

tannergooding replied:
Right, I think this is ultimately in-line with the spec after going through the rest of the code; but it reads a bit funny to me. Not quite sure how to reword it, however....

tannergooding commented on src/jit/codegencommon.cpp:
```diff
@@ -10204,7 +10204,19 @@ var_types Compiler::GetHfaType(CORINFO_CLASS_HANDLE hClass)
+        {
+            // This is a vector type.
+            // HVAs are only supported on ARM64, and only for sizes of 8 or 16 bytes.
+            // For 8-byte vectors corType will be returned as CORINFO_TYPE_DOUBLE.
```
This isn't classified independently as a short vector?

CarolEidt replied:
No; for ABI purposes it is the same (an 8-byte vector is passed in the lower 8-bytes of the register, just as for double)

tannergooding replied:
Is this something that might be useful to differentiate at some point in the future?

CarolEidt replied:
It may be, but it will come at a (probably not large) price in complexity and change to the JIT/EE interface (unless the need for it is internal to the JIT), so I'd propose we wait until such time as we need it.

tannergooding commented on src/jit/compiler.cpp:
```diff
@@ -829,14 +800,21 @@ var_types Compiler::getArgTypeForStruct(CORINFO_CLASS_HANDLE clsHnd,
         {
-            // Structs that are HFA's are passed by value in multiple registers
-            if (IsHfa(clsHnd)
+            // Structs that are HFA's are passed by value in multiple registers.
```
Shouldn't this be `structs that are HFAs/HVAs`?

CarolEidt replied:
I'll update the comment.

tannergooding commented on src/jit/compiler.h:
```diff
@@ -134,6 +134,61 @@ const unsigned FLG_CCTOR = (CORINFO_FLG_CONSTRUCTOR | CORINFO_FLG_STATIC);
+    HFA_ELEM_NONE,
+    HFA_ELEM_FLOAT,
+    HFA_ELEM_DOUBLE,
+    HFA_ELEM_SIMD16
```
Shouldn't this be in a separate `HVA_ELEM` struct that also knows about `SIMD8`?

tannergooding replied:
(this may need to be expanded in the future for float16/float128)

CarolEidt replied:
I'm not sure it's worth the cost of the distinction for `SIMD8` (or `FLOAT16`), as it would be a larger impact to the JIT/EE interface.

tannergooding replied:
I think this is probably fine for now, but it is probably worth distinguishing these long term as it may help diagnosing issues or bringing new platforms onboard.

tannergooding commented on src/jit/compiler.h:
```diff
@@ -1607,29 +1651,33 @@ struct fgArgTabEntry
-                }
+            if (type == TYP_DOUBLE)
+            {
+                // Must be an even number of registers.
```
Why does this this need to be even? Spec says "at most four uniquely addressable members"

CarolEidt replied:
The must be homogeneous, and this is ARM32, where each double is passed in two registers.

tannergooding replied:
Ok, I missed the ARM32 part of this 👍

BruceForstall commented on src/vm/callhelpers.h:
```diff
@@ -39,8 +39,13 @@ struct CallDescrData
 #ifdef ENREGISTERED_RETURNTYPE_MAXSIZE
+#ifdef _TARGET_ARM64_
+    // Use NEON128 to ensure proper alignment for vectors.
+    __declspec(align(16)) NEON128 returnValue[ENREGISTERED_RETURNTYPE_MAXSIZE / sizeof(NEON128)];
```
Should this use the `DECLSPEC_ALIGN(16)` macro?

CarolEidt replied:
I was not aware of it.

CarolEidt replied:
I see that that's what's used in the VM, so I'll change this.

tannergooding commented on src/jit/target.h:
```diff
@@ -246,6 +246,7 @@ typedef unsigned char   regNumberSmall;
   #define FEATURE_MULTIREG_ARGS_OR_RET  1  // Support for passing and/or returning single values in more than one register
   #define FEATURE_MULTIREG_ARGS         0  // Support for passing a single argument in more than one register  
   #define FEATURE_MULTIREG_RET          1  // Support for returning a single value in more than one register
+  #define MAX_PASS_SINGLEREG_BYTES      8  // Maximum size of a struct passed in a single register (double).
```
Shouldn't this be 16 (for SIMD types)?

tannergooding replied:
Oh, I see. This is x86.

CarolEidt replied:
Yes, someday I hope we can split up target.h so that it's not so easy to get confused about which target you're actually looking at!

tannergooding commented on src/jit/target.h:
```diff
@@ -531,6 +532,7 @@ typedef unsigned char   regNumberSmall;
   #define FEATURE_FASTTAILCALL     1       // Tail calls made as epilog+jmp
   #define FEATURE_TAILCALL_OPT     1       // opportunistic Tail calls (i.e. without ".tail" prefix) made as fast tail calls.
   #define FEATURE_SET_FLAGS        0       // Set to true to force the JIT to mark the trees with GTF_SET_FLAGS when the flags need to be set
+  #define MAX_PASS_SINGLEREG_BYTES      8  // Maximum size of a struct passed in a single register (double).
```
Do we have `MAX_RETURN_SINGLEREG_BYTES`? `x64` specifies that `__m128` is returned in register, but passed by ref for the default calling convention.

CarolEidt replied:
We don't have that yet, though presumably it (or something like it) will be added in subsequent ABI cleanups.

tannergooding replied:
Is fixing the x64 ABI for the new vector primtivies on the slate for 3.0?
* Note: I am talking about the x64 default calling convention; not the currently unsupported `__vectorcall` convention

I think the important bits are:
* __m64 types, are passed as if they were integers of the same size.
  * Even if we don't support any intrinsic methods for `Vector64<T>` on the managed side, we should still pass/return it properly.
* For these aggregate types passed as a pointer, including __m128, the caller-allocated temporary memory must be 16-byte aligned.
* A scalar return value that can fit into 64 bits is returned through RAX; this includes __m64 types.
* Non-scalar types including floats, doubles, and vector types such as __m128, __m128i, __m128d are returned in XMM0.

CarolEidt replied:
Short answer: I hadn't planned any changes to the x64 ABI:
- I hadn't realized that `__m64` types were passed in floating point registers - is that type supported/specified for the linux ABI?
- Where does the 16-byte alignment requirement come from? Is that in the ABI spec?

> Non-scalar types including floats, doubles, and vector types such as __m128, __m128i, __m128d are returned in XMM0.

Is this supported without `__vectorcall`?

tannergooding replied:
> is that type supported/specified for the linux ABI

The above are all from the Windows x64 calling convention defined here: https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=vs-2019.

The `__m64`, `__m128`, and `__m256` types are also defined by the Linux "System V" ABI (both 32-bit and 64-bit), but I haven't double checked the requirements for that yet...

> Is this supported without __vectorcall

Yes, as per the above, this is in the default Windows x64 calling convention; so it should be applicable to what we support in the JIT today.

tannergooding replied:
Also, for reference. I am prototyping the ability to marshal blittable generics here: https://github.com/dotnet/coreclr/pull/23899

The Windows x64 failures can be seen in the latest commit and it would likely be useful to validate that the ARM/ARM64 tests pass on that PR after this PR is merged.

tannergooding commented on src/vm/class.cpp:
```diff
@@ -1172,6 +1172,58 @@ bool MethodTable::IsHFA()
+        LPCUTF8 className = GetFullyQualifiedNameInfo(&namespaceName);
+        int vectorSize = 0;
+
+        if (strcmp(className, "Vector`1") == 0)
```
What about `Vector2`, `Vector3`, and `Vector4`? Are they staying as homogeneous aggregates or will they also be treated as `short vector` for the purpose of passing?

CarolEidt replied:
They are explicitly structs of floats, so they will stay has HFAs of float. (That's how I read the ABI spec).

tannergooding replied:
I see your comment about that these are staying as HFAs today.

The upside is that this means it can't be a breaking change; but the downside is that we can't make the existing System.Numerics code less expensive...
* Most of them expose public fields; and something like `Matrix4x4` could have potentially been an HVA, but it will continue being passed `byref`.
* Maybe this can just be resolved with the generic `Vector2<T>`, `Vector3<T>`, and `Vector4<T>` proposal...

CarolEidt replied:
However, unlike the other types, these are existing interop types (and were designed to be), so they need to continue to follow the ARM64 ABI.

CarolEidt commented on src/vm/callingconvention.h:
```diff
@@ -50,8 +50,7 @@ struct ArgLocDesc
 #if defined(_TARGET_ARM64_)
-    bool    m_isSinglePrecision;  // For determining if HFA is single or double
-                                  // precision
+    int      m_hfaFieldSize;      // Size of HFA field
```
Yes, that makes sense.

CarolEidt commented on src/vm/callingconvention.h:
```diff
@@ -589,10 +588,19 @@ class ArgIteratorTemplate : public ARGITERATOR_BASE
+                case ELEMENT_TYPE_R4: hfaFieldSize = 4; break;
+                case ELEMENT_TYPE_R8: hfaFieldSize = 8; break;
+#ifdef _TARGET_ARM64_
+                case ELEMENT_TYPE_VALUETYPE: hfaFieldSize = 16; break;
```
> I guess you're just overloading this existing type for this particular purpose?

Yes, it's a bit of a hack, but avoids having to create a new VM type.

CarolEidt commented on src/vm/class.h:
```diff
@@ -580,7 +587,13 @@ class EEClassLayoutInfo
         {
             LIMITED_METHOD_CONTRACT;
-            m_bFlags |= (hfaType == ELEMENT_TYPE_R4) ? e_R4_HFA : e_R8_HFA;
+            switch (hfaType)
```
I believe it should never be called more than once. I'll add the assert.

CarolEidt commented on src/jit/compiler.h:
```diff
@@ -134,6 +134,61 @@ const unsigned FLG_CCTOR = (CORINFO_FLG_CONSTRUCTOR | CORINFO_FLG_STATIC);
+    HFA_ELEM_NONE,
+    HFA_ELEM_FLOAT,
+    HFA_ELEM_DOUBLE,
+    HFA_ELEM_SIMD16
```
It didn't seem worth ifdef'ing this out

CarolEidt commented on src/jit/compiler.h:
```diff
@@ -666,70 +718,57 @@ class LclVarDsc
+    //
+    // Return Value:
+    //    On Arm64 - Returns 1-4 indicating the number of register slots used by the HFA
+    //    On Arm32 - Returns the total number of single FP register slots used by the HFA, max is 8
```
Yes, I agree, but it's pre-existing and I'm pretty sure that changing it would be non-trivial.

CarolEidt commented on src/jit/compiler.h:
```diff
@@ -666,70 +718,57 @@ class LclVarDsc
+                slots = lvExactSize >> 3;
+                break;
+            case HFA_ELEM_SIMD16:
+                slots = lvExactSize >> 4;
```
That makes sense.

CarolEidt commented:
The "Test pri0 Linux arm64 checked" leg failed with a timeout in /JIT/Methodical/tailcall/Desktop/_il_relthread-race/_il_relthread-race.sh
and "There was an I/O error during the build." on the Windows arm64 leg.

CarolEidt commented:
test Windows_NT arm64 Cross Checked Innerloop Build and Test

CarolEidt commented:
/azp run coreclr-outerloop

CarolEidt commented:
test Windows_NT arm64 Cross Checked Innerloop Build and Test

CarolEidt commented:
@BruceForstall @dotnet/jit-contrib - I've addressed all of the PR feedback, and the only remaining outerloop failures are the recently added test:
JIT/Methodical/largeframes/skip4/skippage4/skippage4.sh

I have the changes from #23881, but this test is a pri1 test and doesn't seem to have been run on that PR (or since). The failure is:
```
Unhandled Exception: System.TypeLoadException: Could not load type 'BigFrames.LargeStructWithRef' from assembly 'skippage4, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' because it contains an object field at offset 65500 that is incorrectly aligned or overlapped by a non-object field.
```
Can I get another review?

CarolEidt commented:
@dotnet/jit-contrib ping

CarolEidt commented:
Thanks @briansull !

echesakovMSFT commented:
@CarolEidt I am seeing 8 R2R Linux arm64 and Linux musl_arm64 tests failing during crossgen run on this commit validation https://dev.azure.com/dnceng/public/_build/results?buildId=157586&view=ms.vss-test-web.build-test-results-tab

CarolEidt commented:
@echesakovMSFT - I'll take a look; are these not run in outerloop testing?

echesakovMSFT commented:
@CarolEidt We don't run R2R tests in **coreclr-outerloop**, only during commit validations and official builds.

@RussKeldorph We probably should add **coreclr-outerloop-r2r** and **coreclr-outerloop-r2r-jistress/gcstress** build definitions for folks who want to trigger those and also have them running on schedule.

CarolEidt commented:
@echesakovMSFT - can you remind me how to invoke a single test with R2R? Is it just the crossgen command with the appropriate "/Platform_Assemblies_Paths" option? Or is there a better way?

RussKeldorph commented:
@CarolEidt To crossgen the test, set `RunCrossGen=1` in the environment before running the test script.  If you need the framework compiled like the R2R jobs, pass the `crossgen` option to build-test's command line.
